### PR TITLE
support for syncing a shared DH keyring down from the server

### PR DESCRIPTION
**Pull request by maxtaco**
- we don't want these keys to ever make it out to disk, so let's just keep them in memory
- check that they are encrypted from authorized encryptors, and also that they match the public keys advertised in the sigchain
- untested just yet

### CODE REVIEW

**Comment by maxtaco** [comment]
@mlsteele this is blocked on your PR for testing purposes. But the code is all there, I think...

**Comment by mlsteele** [comment]
put up https://github.com/keybase/client/pull/6595
you can reset your `pvl-tools/kit.json`.

File: go/libkb/keyring.go [hunk 1]
```diff
@@ -24,14 +24,7 @@ type KeyringFile struct {
...
 	Contextified
-}
-
-func (k *Keyrings) MakeKeyrings(filenames []string, isPublic bool) []*KeyringFile {
-	v := make([]*KeyringFile, len(filenames), len(filenames))
-	for i, filename := range filenames {
-		v[i] = &KeyringFile{filename, openpgp.EntityList{}, isPublic, nil, nil, Contextified{g: k.G()}}
-	}
-	return v
+	SharedDHKeyring SharedDHKeyring
```
Comment (mlsteele) [inline go/libkb/keyring.go]: This is left uninitialized on Gs. Is that on purpose?

File: go/libkb/id_table.go [hunk 2]
```diff
@@ -721,6 +721,40 @@ func (s *SubkeyChainLink) insertIntoTable(tab *IdentityTable) {
...
+func ParseSharedDHKeyChainLink(b GenericChainLink) (ret *SharedDHKeyChainLink, err error) {
+	var kid, pkid keybase1.KID
+	var g int
+	section := b.payloadJSON.AtPath("body.shared_dh_key")
+	if kid, err = GetKID(section.AtKey("kid")); err != nil {
+		err = ChainLinkError{fmt.Sprintf("Can't get KID for shared_dh @%s: %s", b.ToDebugString(), err)}
+	} else if g, err = section.AtKey("generation").GetInt(); err != nil {
+		err = ChainLinkError{fmt.Sprintf("Can't get generation for shared_dh @%s: %s", b.ToDebugString(), err)}
+	} else {
+		ret = &SharedDHKeyChainLink{b, kid, pkid, SharedDHKeyGeneration(g)}
```
Comment (mlsteele) [inline go/libkb/id_table.go]: `pkid` is always blank. What is `parentKid` anyway?

File: go/libkb/shared_dh.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+		Contextified: NewContextified(g),
+		uid:          uid,
+		generations:  make(SharedDHKeyMap),
+	}
+}
+
+// CurrentGeneration returns what generation we're on. The version possible
+// Version is 1. Version 0 implies no keys are available.
+func (s *SharedDHKeyring) CurrentGeneration() SharedDHKeyGeneration {
+	return SharedDHKeyGeneration(len(s.generations))
```
Comment (mlsteele) [inline go/libkb/shared_dh.go]: Needs a lock to be public. But also needs to not deadlock inside other `*Locked` functions, so could add `currentGenerationLocked`.

File: go/libkb/shared_dh.go [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+}
+
+func (s *SharedDHKeyring) SharedDHKey(g SharedDHKeyGeneration) *NaclDHKeyPair {
+	key, found := s.generations[g]
```
Comment (mlsteele) [inline go/libkb/shared_dh.go]: Needs a lock.

File: go/libkb/shared_dh.go [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+		},
+		SessionType: APISessionTypeREQUIRED,
+		RetryCount:  5, // It's pretty bad to fail this, so retry.
+		NetContext:  ctx,
+	})
+	if err != nil {
+		return nil, err
+	}
+	var boxes sharedDHSecretKeyBoxes
+	if err = res.Body.UnmarshalAgain(&boxes); err != nil {
```
Comment (mlsteele) [inline go/libkb/shared_dh.go]: `GetDecode`? I don't feel strongly either way

File: go/libkb/keyring.go (hunk 1 above)
Comment (maxtaco) [inline go/libkb/keyring.go]: yeah, it is for now, I wasn't sure how we wanted to hook it in.

File: go/libkb/id_table.go (hunk 2 above)
Comment (maxtaco) [inline go/libkb/id_table.go]: remnant. will nuke it.  Only makes sense for subkeys

File: go/libkb/shared_dh.go (hunk 3 above)
Comment (maxtaco) [inline go/libkb/shared_dh.go]: thx!

File: go/libkb/shared_dh.go (hunk 4 above)
Comment (maxtaco) [inline go/libkb/shared_dh.go]: thx!
willfix

File: go/libkb/shared_dh.go (hunk 5 above)
Comment (maxtaco) [inline go/libkb/shared_dh.go]: good idea

**Comment by maxtaco** [comment]
ok all comments addressed, except for the first

**Comment by maxtaco** [comment]
Actually addressed that too, it's no longer needed.

File: go/libkb/keyring.go (hunk 1 above)
Comment (maxtaco) [inline go/libkb/keyring.go]: this was ill-conceived, nuked...

**Comment by maxtaco** [comment]
dropped `pvl/kit.json`

**Pull request merged**